### ISSUE TITLE
Support statically analyzable singletons

### ISSUE THREAD

**Issue by alippai** [microsoft/pyright#8764]
While I understand that using classes can be a workaround, but it's super verbose.
Supporting something like this would be amazing:
```python
from types import SimpleNamespace
class Table(SimpleNamespace):
  pass

users_table = Table(id=Col('INT'), name=Col('VARCHAR'))
posts_table = Table(user_id=Col('INT'), timestamp=Col(TIMESTAMP'), post_id=Col('INT'), message=Col('VARCHAR'))
comments_table = Table(user_id=Col('INT'), timestamp=Col(TIMESTAMP'), post_id=Col('INT'), comment_id=Col('INT'), message=Col('VARCHAR'))
```

Right now I have to:
```python
class UsersTable:
  id: Col
  name: Col

class PostsTable:
  user_id: Col
  timestamp: Col
  post_id: Col
  message: Col

class CommentsTable:
  user_id: Col
  timestamp: Col
  post_id: Col
  comment_id: Col
  
users_table = UsersTable()
posts_table = PostsTable()
comments_table = CommentsTable()
```

I don't know if this is expressible with python types (I don't know a way to describe the types of individual keys using typing), however technically the `SimpleNamespace` or some magic protocol could be supported.

**Comment by erictraut** [comment]
I'm not sure what you mean. Your code sample above has syntax errors and is incomplete (undefined symbols). If my understanding is correct, what you're proposing would require an update to the Python typing spec. If you'd like to pursue that, I recommend posting your idea to the [Python typing forum](https://discuss.python.org/c/typing/32) to get feedback from the broader typing community. [Here](https://github.com/python/typing-council#decisions) is the full process for proposing updates to the typing spec. Depending on what you're proposing here, it may also require a formal PEP.

**Comment by alippai** [comment]
Apologies if my example made it harder to understand the suggestion instead of making it more transparent.

In short I propose supporting the special `SimpleNamespace` as the keys and types can be inferred (if it's not dynamic).

I understand `SimpleNamespace` doesn't have a Python typing definition.
As far as I know what it does is not expressible without a new typing spec and you are right suggesting that it could be improved upstream.

Since it's a python built-in, pyright might add support to it anyways (without having a PIP and it being expressed using python typing rules).

```
from types import SimpleNamespace
my_obj = SimpleNamespace(A=1, B='abc')
my_obj.B # pyright could infer myobj has B with the type of str, currently it's Any
```

**Comment by erictraut** [comment]
`SimpleNamespace` supports arbitrary keys. [Here](https://github.com/python/typeshed/blob/1ace5718deaf3041f8e3d1dc9c9e8a8e830e517f/stdlib/types.pyi#L317) is how it is defined in the typeshed stubs. Pyright is honoring that type definition.

`SimpleNamespace` is not a good choice if you want your code to be type checked. A `dataclass` would be a better option. That way, you can declare the fields that are present and their corresponding types.

**Comment by alippai** [comment]
Yes, arbitrary keys known ahead without code evaluation required. It’s out of scope for pyright, understood. Thanks for the details and the pointers.

**Comment by erictraut** [comment]
As a general rule, pyright does not override type information in typeshed unless the typing spec says that type checkers should special-case the behavior. I've made a few exceptions to this rule over the years, and I've come to regret most of these decisions because it leads to problems when the typeshed definitions are updated.

Let's see what others in the typing forum think about special-casing `SimpleNamespace`. I'm guessing that the response to this idea will be somewhere between lukewarm and negative. In my mind, `SimpleNamespace` is the antithesis of static typing. It's a completely dynamic type that allows for arbitrary attributes with arbitrary value types. If you want to use static type checking in your code, there are many better alternatives such as dataclasses or named tuples.

**Comment by alippai** [comment]
I love pyright and your decisions (like the above) made it great. You have to draw a line somewhere and I trust you better than myself on this topic.
